gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ImmunizationMapper.java

```diff
@@ -0,0 +1,248 @@
+    }
+
+    private Reference getEncounterReference(List<Encounter> encounterList, II ehrCompositionId) {
+        if (ehrCompositionId != null) {
+            var matchingEncounter = encounterList.stream()
+                .filter(encounter -> matchingIds(encounter.getId(), ehrCompositionId))
+                .findFirst();
+
+            if (matchingEncounter.isPresent()) {
+                return new Reference(matchingEncounter.get().getIdElement());
+            }
+        }
+
+        return null;
+    }
+
+    private boolean matchingIds(String encounterId, II ehrCompositionId) {
+        return encounterId.equals(ehrCompositionId.getRoot());
+    }
+
+    private List<RCMRMT030101UK04ObservationStatement> getImmunizationObservationStatements(RCMRMT030101UK04EhrComposition ehrComposition) {
+        return ehrComposition.getComponent()
+            .stream()
+            .map(RCMRMT030101UK04Component4::getObservationStatement)
+            .filter(this::hasImmunizationCode)
+            .collect(Collectors.toList());
+    }
+
+    private Extension createRecordedTimeExtension(RCMRMT030101UK04EhrComposition ehrComposition) {
+        if (ehrComposition.getAuthor() != null) {
+            return new Extension()
+                .setValue(new StringType(ehrComposition.getAuthor().getTime().getValue()));
+        } else if (ehrComposition.getEffectiveTime() != null) {
+            if (ehrComposition.getAvailabilityTime().getNullFlavor() == null) {
+                return new Extension()
+                    .setValue(new StringType(DateFormatUtil.parseToDateTimeType(ehrComposition.getAvailabilityTime().getValue())
+                        .asStringValue()));
+            }
+        }
+
+        return null;
+    }
+
+    private Identifier getIdentifier(String id) {
+        return new Identifier()
+            .setSystem(IDENTIFIER_SYSTEM) // TODO: concatenate source practice org id to URL (NIAD-2021)
+            .setValue(id);
+    }
+
+    private boolean hasImmunizationCode(RCMRMT030101UK04ObservationStatement observationStatement) {
+        String snomedCode = observationStatement.getCode().getCodeSystem();
+
+        return IMMUNIZATION_SNOMED_CODE.equals(snomedCode);
+    }
+
+    private Extension createVaccineProcedureExtension(RCMRMT030101UK04ObservationStatement observationStatement) {
+        return new Extension()
+            .setUrl(VACCINE_PROCEDURE_URL)
+            .setValue(new Extension()
+                .setUrl(VACCINATION_CODING_EXTENSION_URL)
+                .setValue(codeableConceptMapper.mapToCodeableConcept(observationStatement.getCode()))
+            );
+    }
+
+    private String getTSStringValue(TS ts) {
+        if (ts == null) {
+            return null;
+        } else if (ts.getValue() != null) {
+            return ts.getValue();
+        } else if (ts.getNullFlavor().equals(CsNullFlavor.UNK)) {
+            return CsNullFlavor.UNK.value();
+        }
+
+        return null;
```

Comment:
You could simplify it by checking if ts is not null first, then you'll only have to return null once:
```suggestion
        if (ts != null) {
            if (ts.getValue() != null) {
                return ts.getValue();
            } else if (ts.getNullFlavor().equals(CsNullFlavor.UNK)) {
                return CsNullFlavor.UNK.value();
            }
        }

        return null;
```